sn_networking/src/metrics/bad_node.rs

```diff
@@ -77,38 +111,158 @@ impl TimeFrameType {
+    ) -> tokio::sync::mpsc::Sender<BadNodeMetricsMsg> {
+        let mut bad_node_metrics = BadNodeMetrics {
+            shunned_count_across_time_frames: ShunnedCountAcrossTimeFrames {
+                metric: time_based_shunned_count,
+                shunned_report_tracker: Vec::new(),
+            },
+            shunned_by_close_group: ShunnedByCloseGroup {
+                metric_current_group: shunned_by_close_group,
+                metric_old_group: shunned_by_old_close_group,
 
+                close_group_peers: Vec::new(),
+                old_close_group_peers: Vec::new(),
+                old_close_group_peers_that_have_shunned_us: HashSet::new(),
+                close_group_peers_that_have_shunned_us: HashSet::new(),
+            },
+        };
+
+        let (tx, mut rx) = tokio::sync::mpsc::channel(10);
         tokio::spawn(async move {
-            let mut shunned_metrics = ShunnedCountAcrossTimeFrames {
-                metric: time_based_shunned_count,
-                tracked_values: Vec::new(),
-            };
             let mut update_interval = interval(UPDATE_INTERVAL);
             update_interval.tick().await;
 
             loop {
                 tokio::select! {
-                    _ = rx.recv() => {
-                        shunned_metrics.record_shunned_metric();
+                    msg = rx.recv() => {
+                        match msg {
+                            Some(BadNodeMetricsMsg::ShunnedByPeer(peer)) => {
+                                bad_node_metrics.shunned_count_across_time_frames.record_shunned_metric();
+                                bad_node_metrics.shunned_by_close_group.record_shunned_metric(peer);
+
+                            }
+                            Some(BadNodeMetricsMsg::CloseGroupUpdated(new_closest_peers)) => {
+                                bad_node_metrics.shunned_by_close_group.update_close_group_peers(new_closest_peers);
+                            }
+                            None => break,
+                        }
+
 
                     }
                     _ = update_interval.tick() => {
-                        shunned_metrics.update();
+                        bad_node_metrics.shunned_count_across_time_frames.try_update_state();
                     }
                 }
             }
         });
         tx
     }
+}
+
+impl ShunnedByCloseGroup {
+    pub(crate) fn record_shunned_metric(&mut self, peer: PeerId) {
+        // increment the metric if the peer is in the close group (new or old) and hasn't shunned us before
+        if self.close_group_peers.contains(&peer) {
+            if !self.close_group_peers_that_have_shunned_us.contains(&peer) {
+                self.metric_current_group.inc();
+                self.close_group_peers_that_have_shunned_us.insert(peer);
+            }
+        } else if self.old_close_group_peers.iter().any(|(p, _)| p == &peer)
+            && !self
+                .old_close_group_peers_that_have_shunned_us
```

Comment:
looks like `close_group_peers_that_have_shunned_us` and `old_close_group_peers_that_have_shunned_us` can be merged into one set of `peers_shunned_us`

Then there are couple of places that can have code simplified to avoid `shifting between these two sets` .

Reply:
Hey @maqi, we already have a `shunned_count` metric available to us, but it does not paint the whole picture as we care more about if our close group still considers us as bad.
These metrics are primarily focused on LP to show the node's health.

We need these 2 metrics as separate ones as our close group will keep changing once we are marked as bad by them (as they block communication to us, but not instantly as we still might keep them in our RT for a while), so we need both of them.

Reply:
i think the separated metric you want is : `self.metric_current_group;` and `self.metric_old_group` ?
I'm not saying merge these two.

Say if there is merged `peers_shunned_us`, then the code block here will be:
```rust
if !self.peers_shunned_us.contains(&peer){
    if self.close_group_peers.contains(&peer) {
        self.metric_current_group.inc();
    } else if self.old_close_group_peers.iter().any(|(p, _)| p == &peer) {
        self.metric_old_group.inc();
    }    
}
let _ = self.peers_shunned_us.insert(peer);
```

Reply:
Right rigt I see.. Lemme try it out! That will simplify the code a lot.

Reply:
And you can then have third metric: `non-close-group-peers that shunned us`
which is: `self.peers_shunned_us.len() - self.metric_current_group - self.metric_old_group`